### PR TITLE
Ripsaw CLI

### PR DESCRIPTION
**Pull request by whitleykeith**
### Description

Adds `ripsaw-cli` python project to the repo that is meant to be a lightweight cli tool used to install the operator and run benchmarks.

### CODE REVIEW

File: cli/ripsaw/clients/k8s.py
Comment (jtaleric) [inline cli/ripsaw/clients/k8s.py:1]: Missing License ;)

File: cli/ripsaw/commands/benchmark.py
Comment (jtaleric) [inline cli/ripsaw/commands/benchmark.py:1]: License

File: cli/ripsaw/commands/operator.py
Comment (jtaleric) [inline cli/ripsaw/commands/operator.py:1]: License

File: cli/ripsaw/clients/k8s.py
Comment (whitleykeith) [reply cli/ripsaw/clients/k8s.py:15]: added

File: .pre-commit-config.yaml
Comment (learnitall) [inline .pre-commit-config.yaml:63]: I think that this exclude might be leftover from snafu's pre-commit config? Might be ok to chuck it

File: .github/workflows/cli.lint.yml
Comment (learnitall) [inline .github/workflows/cli.lint.yml:1]: If we want to use GHA for pre-commit rather than the pre-commit CI (https://pre-commit.ci/) than we should remove the CI configuration from the pre-commit config. I personally would prefer the pre-commit CI because it's really useful, and it's the direction I want to take benchmark-wrapper, but up to y'all.

File: .pylintrc
Comment (learnitall) [inline .pylintrc:4]: This seems to also be leftover from snafu, could remove this ignore path.

File: cli/.coveragerc
Comment (learnitall) [inline cli/.coveragerc:4]: Why are we ignoring the CLI and ``commands/`` sub-dir? Are those out of scope for the unit tests?

File: cli/ripsaw/clients/k8s.py
Comment (learnitall) [inline cli/ripsaw/clients/k8s.py:96]: This should be an easy fix that doesn't require an ignore? By replacing the below brackets with parens?

```python
any((pod.status.phase != "Running" for pod in pods))
```

File: cli/setup.cfg
Comment (learnitall) [inline cli/setup.cfg:35]: Can we add these on separate lines to match the format from above?

File: cli/ripsaw/util/exceptions.py
Comment (learnitall) [inline cli/ripsaw/util/exceptions.py:20]: Is there a reason why this was disabled rather than changing the line to the following?

```python
super().__init__(msg)
```

**Review by learnitall — APPROVED** [review]
Really well done, looks great! Got my LGTM. I left a couple of nitpicky comments, but other than that this is great. Easy to read and understand.

File: .github/workflows/cli.lint.yml
Comment (whitleykeith) [reply .github/workflows/cli.lint.yml:1]: Does pre-commit CI require each dev to install/setup pre-commit on their machines? TBH I want both. I want the pre-commit hooks so that if I enable it I can catch my lint errors earlier, but I want to ensure that every PR gets linted anyways

File: .pylintrc
Comment (whitleykeith) [reply .pylintrc:4]: same thing as the problem with the bandit exclusions. The cli testsuses fixtures which trigger `unused-argument` or `unsued-import` if the fixture is ffrom a different module. Instead of littering all the tests with pylint disables I just excluded the tests from linting since I didn't have much luck catching many issues there with them

File: .pre-commit-config.yaml
Comment (whitleykeith) [reply .pre-commit-config.yaml:63]: The tests uses asserts which made bandit fail but I didn't want to remove skip the rule entirely since I want it to fail if asserts are in the actual code base

File: cli/.coveragerc
Comment (whitleykeith) [reply cli/.coveragerc:4]: for unit tests, yeah. The `cli.py` file is just the code entrypoint and only really executes if you run the cli itself. `commands` I will remove from there because even though it's out of scope for unit tests we shouldn't ignore it from coverage

File: cli/ripsaw/util/exceptions.py
Comment (whitleykeith) [reply cli/ripsaw/util/exceptions.py]: fixed

File: cli/setup.cfg
Comment (whitleykeith) [reply cli/setup.cfg]: fixed

File: cli/ripsaw/clients/k8s.py
Comment (whitleykeith) [reply cli/ripsaw/clients/k8s.py]: fixed

File: .pre-commit-config.yaml
Comment (learnitall) [reply .pre-commit-config.yaml:65]: My bad on these `cli/tests` path comments that I left. I need to go get a pair of reading glasses or something- I mistook that for `ci/tests`. Please ignore, my bad :+1:

File: cli/.coveragerc
Comment (learnitall) [reply cli/.coveragerc:4]: Sounds good! Thanks for the clarification.

File: .github/workflows/cli.lint.yml
Comment (learnitall) [reply .github/workflows/cli.lint.yml:1]: Yep, yeah the only thing that the pre-commit CI will do is replace the Github Workflow. All devs should still install, setup and use pre-commit. The pre-commit CI will run on each push and on each sync of a pull-request, auto-fix pull-requests, create PRs that auto-fix code when possible, and create PRs to upgrade hook versions when possible. Additionally it's really fast because they do a lot of install caching. 

And to clarify, by "auto-fix" I mean that if a hook modifies files during its execution, pre-commit will take those modifications and submit a PR for them.

File: config/manager/kustomization.yaml
Comment (rsevilla87) [inline config/manager/kustomization.yaml:16]: I think you uploaded this file by mistake, (It gets modified whenever you make deploy)

File: .github/workflows/cli.lint.yml
Comment (whitleykeith) [reply .github/workflows/cli.lint.yml:1]: Could we use the pre-commit CI to fix the PR before merging? I really just want PRs to not get merged unless they've been linted. If we can do that then i'm fine with not using GHA

File: config/manager/kustomization.yaml
Comment (whitleykeith) [reply config/manager/kustomization.yaml]: fixed

File: cli/ripsaw/commands/benchmark.py
Comment (sjug) [inline cli/ripsaw/commands/benchmark.py:28]: Is this just placeholder text? (x3)

File: cli/ripsaw/commands/operator.py
Comment (sjug) [inline cli/ripsaw/commands/operator.py:72]: Does this override the default repo/branch based on the git repo & branch if you're in one locally as it's not reflected in that initial "Installing Operator from repo..." print statement?

Comment (sjug) [reply cli/ripsaw/commands/operator.py:72]: When not installing/deleting outside of a git repo it's not too happy:

```
$ ripsaw operator install
Installing Operator from repo https://github.com/cloud-bulldozer/benchmark-operator and branch master
Traceback (most recent call last):
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/bin/ripsaw", line 33, in <module>
    sys.exit(load_entry_point('ripsaw-cli==0.0.1', 'console_scripts', 'ripsaw')())
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/lib/python3.9/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/lib/python3.9/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/lib/python3.9/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/lib/python3.9/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/lib/python3.9/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/lib/python3.9/site-packages/ripsaw/commands/operator.py", line 42, in install
    _perform_operator_action(repo, branch, command, kubeconfig)
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/lib/python3.9/site-packages/ripsaw/commands/operator.py", line 72, in _perform_operator_action
    if "benchmark-operator" in local_git_repo:
TypeError: argument of type 'NoneType' is not iterable
```

**Review by dry923 — APPROVED** [review]
LGTM!

File: cli/ripsaw/commands/operator.py
Comment (sjug) [inline cli/ripsaw/commands/operator.py:75]: Can we handle `CalledProcessError` for this?

File: cli/ripsaw/commands/benchmark.py
Comment (whitleykeith) [reply cli/ripsaw/commands/benchmark.py:28]: This function leverages a feature of Click CLI to "group" commands which creates a decorator you can use (see L31). The docstring is really just there to explain why this function has to exist.

File: cli/ripsaw/commands/operator.py
Comment (whitleykeith) [reply cli/ripsaw/commands/operator.py]: yep the conditional needed to check to make sure `local_git_repo is not None`. Just pushed a fix

Also fixed the logs around telling you what repo/branch you're on.

File: cli/ripsaw/commands/operator.py
Comment (whitleykeith) [reply cli/ripsaw/commands/operator.py:75]: yeah we can, what do you think we should do with it? log/reraise/exit?

File: cli/ripsaw/commands/operator.py
Comment (sjug) [reply cli/ripsaw/commands/operator.py]: In the case where the user provides `ripsaw operator install` a `--repo` or `--branch` inside of the git repo, can we have the CLI args override even the git repo?

File: cli/ripsaw/commands/operator.py
Comment (sjug) [reply cli/ripsaw/commands/operator.py:75]: Whatever you prefer, the stack trace doesn't look like an expected behavior.

File: cli/ripsaw/commands/benchmark.py
Comment (sjug) [reply cli/ripsaw/commands/benchmark.py:28]: Right, I understand that. Why don't we use that space to explain why this function has to exist to make this decorator string useful?

File: cli/ripsaw/commands/operator.py
Comment (sjug) [inline cli/ripsaw/commands/operator.py:87]:
```
  File "/home/user/go/src/github.com/cloud-bulldozer/benchmark-operator/venv/lib/python3.9/site-packages/ripsaw/commands/operator.py", line 87, in _perform_operator_action
    _clone_repo(repo, branch, temp_dir.name)
AttributeError: 'str' object has no attribute 'name'
```

File: cli/ripsaw/commands/operator.py
Comment (whitleykeith) [reply cli/ripsaw/commands/operator.py]: the only real way to do that is to add a flag to force using a remote repo, which I just added as `--force-remote`. This has the added benefit of being able to force using the remote repo with the default repo/branch

File: cli/ripsaw/commands/operator.py
Comment (whitleykeith) [reply cli/ripsaw/commands/operator.py]: updated it to make it look cleaner and also refactored it a bit

File: cli/ripsaw/commands/operator.py
Comment (whitleykeith) [reply cli/ripsaw/commands/operator.py]: fixed

**Pull request merged**
